matRad_DijSampling.m

```diff
@@ -35,25 +34,63 @@
+ixSampTail          = ix(linIxSample);                           % indices that are going to be reduced by sampling
+
+%% sample for each radiological depth the lateral halo dose  
+radDepthTail        = (radDepthV(linIxSample));                  % get radiological depth in the tail
+
+% cluster radiological dephts to reduce computations
+B_r                 = int32(ceil(radDepthTail));                 % cluster radiological depths; hist(B,NumOfClusters) could also be used
+if boolFineClustering 
+    [C,~,~]      = unique(B_r);                                                         % get unique radiological depht values == fine clustering
+else 
+     maxRadDepth = double(max(B_r));
+     C           = int32(linspace(0,maxRadDepth,round(maxRadDepth)/deltaRadDepth));     % coarse clustering of rad depths    
 end
 
-maxDose            = max(bixelDose);
-linIxSample        = find(bixelDose < maxDose * relDoseLimits(1) & bixelDose > maxDose * relDoseLimits(2));
-NumSamples         = fix(numel(linIxSample)*SamplingRate);
-bixelSampDose      = bixelDose(linIxSample);
-Prob               = bixelSampDose/max(bixelSampDose);
-ProbNorm           = sort(Prob,'descend')./(sum(Prob));
-CDF                = cumsum(ProbNorm);
-randomValues       = (CDF(end)-CDF(1)).* rand(NumSamples,1) + CDF(1);
-ixSamp             = interp1(CDF,linIxSample,randomValues,'nearest');
-ixNew              = (bixelDose > maxDose * relDoseLimits(1)) ;
-ixNew(ixSamp)      = 1;
-bixelDose(ixSamp)  = maxDose * relDoseLimits(1);
-bixelDose          = bixelDose(ixNew);
-ix                 = ix(ixNew);
+ixNew               = zeros(numTail,1);                          % inizialize new index vector
+bixelDoseNew        = zeros(numTail,1);                          % inizialize new dose vector
+IxCnt               = 1;
+linIx               = int32(1:1:numTail)';
+
+% loop over clustered radiological depths
+for i = 1:numel(C)-1
 
+    ixTmp              = linIx(B_r >= C(i) & B_r < C(i+1));      % extracting sub indices
+    if isempty(ixTmp)
+        continue
+    end 
+    subDose            = bixelDoseTail(ixTmp);                   % get tail dose in current cluster
+    subIx              = ixSampTail(ixTmp);                      % get indices in current cluster
+    
+    thresholdDose      = max(subDose); 
+    r                  = rand(numel(subDose),1);                 % get random samples
+    ixSamp             = r<=(subDose/thresholdDose);
+    NumSamples         = sum(ixSamp);
 
+    ixNew(IxCnt:IxCnt+NumSamples-1,1)        = subIx(ixSamp);    % save new indices
+    bixelDoseNew(IxCnt:IxCnt+NumSamples-1,1) = thresholdDose;    % set the dose
+    
+    IxCnt = IxCnt + NumSamples;
+end
 
+% cut the vectors
+ixNew        = ixNew(1:IxCnt-1);
+bixelDoseNew = bixelDoseNew(1:IxCnt-1);
+% add inner core values  
+ixNew        = [ix(ixCore); ixNew];
+bixelDoseNew = [bixelDoseCore; bixelDoseNew];
 
+end
```

Comment:
Two general comments:

1. What is the time consuming part of this function now? If there is an issue with all the index administration we could certainly think about alternative strategies. E.g. you could sequentially delete and replace values from the index and dose vector making a lot of book keeping superfluous.

2. In principle this function could also be used for particles - right? But for particles you probably want the sampling set to be defined globally by the dose (e.g. 1% of the max) and not by a constant radius.